### PR TITLE
cpuid crate refactoring - part 2

### PR DESCRIPTION
**Pull request by serban300**
Signed-off-by: Serban Iorga <seriorga@amazon.com>

Issue #, if available: #851 

Description of changes:
    - move leaf handlers into separate functions
    - test leaf handlers individually
    - move all the templates into a folder

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cpuid/src/cpu_leaf/leaf_0x1/engine.rs [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+    let mut max_addressable_lcpu = (cpu_count as f64).log2().ceil();
+    max_addressable_lcpu = (2 as f64).powf(max_addressable_lcpu);
+    // check that this number is still an u8
+    if max_addressable_lcpu > u8::max_value().into() {
+        return Err(Error::VcpuCountOverflow);
+    }
+    Ok(max_addressable_lcpu as u8)
+}
+
+pub struct Engine {
```
Comment (andreeaflorescu) [inline cpuid/src/cpu_leaf/leaf_0x1/engine.rs]: What is Engine suppose to be?

Comment (serban300) [reply cpuid/src/cpu_leaf/leaf_0x1/engine.rs]: LeafEngine.process is supposed to be the handler that configures each leaf. There was a switch-case in filter_cpuid that used to do that.

Engine is just an implementation of LeafEngine

**Comment by serban300** [comment]
> I don't think it is a good idea to have one module for each leaf. The code it is too sparse and hard to follow. What is the benefit of that? I think it is good practice to explain the **why** in PRs for cases where the need for the change is not obvious.
> 
> If you create a template module, there is no need to name the individual files c3_template and t2_template. You can name them just t2 and c3 as they are going to be imported as use cpuid::template::{c3, t2}.

I agree with naming the template files `t2` and `c3`

By splitting the cpu_leaf logic I wanted to avoid having a file that is bigger then 1000 lines and that contains multiple test modules (one per LeafEngine) and uses multiple imbrication layers. I tried to do it like that and I couldn't follow the code. Also in the future I would like to maybe be able to split the leafs into multiple folders by cpu vendor if needed. For example:

- common
    - leaf_bstr.rs
- intel
    - leaf_0x4.rs
- amd
    - leaf_0x8000001d.rs

In general I think that modularity is an asset, not a liability. Anyway, let's discuss this offline.

File: cpuid/src/transformer/common.rs [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    fn test() {
+        assert_eq!(get_max_cpus_per_package(1).unwrap(), 1);
+        assert_eq!(get_max_cpus_per_package(2).unwrap(), 2);
+        assert_eq!(get_max_cpus_per_package(4).unwrap(), 4);
+        assert_eq!(get_max_cpus_per_package(6).unwrap(), 8);
+
+        assert!(get_max_cpus_per_package(u8::max_value()).is_err());
+    }
+
+    fn test_update_feature_info_entry(cpu_count: u8, expected_htt: bool) {
```
Comment (andreeaflorescu) [inline cpuid/src/transformer/common.rs]: nit: you can name this check_update.... because we are prefixing with test on the test functions.
Same for the other test helper functions.

Comment (serban300) [reply cpuid/src/transformer/common.rs]: Done

**Comment by dianpopa** [comment]
@serban300 It would be a lot easier to review if the changes got split into multiple commits. I see that in the **Desctiption of changes** from the conversation section you are enumerating three directions this PR heads towards. Would it be feasible to do a commit for each of them?

**Comment by serban300** [comment]
> @serban300 It would be a lot easier to review if the changes got split into multiple commits. I see that in the **Desctiption of changes** from the conversation section you are enumerating three directions this PR heads towards. Would it be feasible to do a commit for each of them?

The first 2 points are connected so I don't think it's feasible to split them into separate commits.
I could try to split the 3rd point into a separate commit, but I don't know if it would make the review much easier since that change only involves moving 2 files into a subfolder.

File: cpuid/src/cpu_leaf.rs [hunk 3]
```diff
@@ -81,11 +81,11 @@ pub mod leaf_0x4 {
             msb_index: 7,
             lsb_index: 5,
         };
-        pub const MAX_ADDR_IDS_SHARING_CACHE_BITRANGE: BitRange = BitRange {
+        pub const MAX_CPUS_PER_CORE: BitRange = BitRange {
```
[the remark this answers is not in the document]
Comment (dhrgit) [reply cpuid/src/cpu_leaf.rs]: Oh, right - the const / function thingy flew right past me. Anyway, what I was getting at was a clear / easy / quick way to scan the source and see what those constants meant.

**Comment by serban300** [comment]
@dhrgit I addressed all the comments. Please take another look.
